flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java

```diff
@@ -0,0 +1,466 @@
+				LOG.debug("Timed out polling events from network, reacquiring subscription - {} ({})", shardId, consumerArn);
+				return false;
+			} else if (subscriptionEvent.isSubscribeToShardEvent()) {
+				SubscribeToShardEvent event = subscriptionEvent.getSubscribeToShardEvent();
+				continuationSequenceNumber = event.continuationSequenceNumber();
+				if (!event.records().isEmpty()) {
+					eventConsumer.accept(event);
+				}
+			} else if (subscriptionEvent.isSubscriptionComplete()) {
+				// The subscription is complete, but the shard might not be, so we return incomplete
+				return false;
+			} else {
+				handleError(subscriptionEvent.getThrowable());
+				return false;
+			}
+		} while (continuationSequenceNumber != null);
+
+		return true;
+	}
+
+	/**
+	 * The {@link FanOutShardSubscription} subscribes to the events coming from KDS and adds them to the {@link BlockingQueue}.
+	 * Backpressure is applied based on the maximum capacity of the queue.
+	 * The {@link Subscriber} methods of this class are invoked by a thread from the {@link KinesisAsyncClient}.
+	 */
+	private class FanOutShardSubscription implements Subscriber<SubscribeToShardEventStream> {
+
+		private Subscription subscription;
+
+		private boolean cancelled = false;
+
+		private final CountDownLatch waitForSubscriptionLatch;
+
+		private final Object lockObject = new Object();
+
+		private FanOutShardSubscription(final CountDownLatch waitForSubscriptionLatch) {
+			this.waitForSubscriptionLatch = waitForSubscriptionLatch;
+		}
+
+		/**
+		 * Flag to the producer that we are ready to receive more events.
+		 */
+		public void requestRecord() {
+			if (!cancelled) {
+				LOG.debug("Requesting more records from EFO subscription - {} ({})", shardId, consumerArn);
+				subscription.request(1);
+			}
+		}
+
+		@Override
+		public void onSubscribe(Subscription subscription) {
+			this.subscription = subscription;
+			waitForSubscriptionLatch.countDown();
+		}
+
+		@Override
+		public void onNext(SubscribeToShardEventStream subscribeToShardEventStream) {
+			subscribeToShardEventStream.accept(new SubscribeToShardResponseHandler.Visitor() {
+				@Override
+				public void visit(SubscribeToShardEvent event) {
+					synchronized (lockObject) {
+						if (enqueueEventWithRetry(new SubscriptionNextEvent(event))) {
+							requestRecord();
+						}
+					}
+				}
+			});
+		}
+
+		@Override
+		public void onError(Throwable throwable) {
+			LOG.debug("Error occurred on EFO subscription: {} - ({}).  {} ({})",
+				throwable.getClass().getName(), throwable.getMessage(), shardId, consumerArn);
+
+			// Cancel the subscription to signal the onNext to stop queuing and requesting data
+			cancelSubscription();
+
+			synchronized (lockObject) {
+				// Empty the queue and add a poison pill to terminate this subscriber
+				// The synchronized block ensures that new data is not written in the meantime
+				queue.clear();
+				enqueueEvent(new SubscriptionErrorEvent(throwable));
```

Comment:
In the case of backpressure in Flink's network stack, would this also mean that job failure is postponed until this `SubscriptionErrorEvent` is postponed until the source connector thread consumes it?

Reply:
Yes. Most errors are recoverable. But if a non-recoverable error is thrown and application is currently consuming a record batch. The application would finish processing the record batch, then receive the error and cancel the Flink job. I think this is ok, do you have concerns?

In the case of multiple shards, typically a non-recoverable error from the network stack would be detected by all `ShardConsumers` and it would be a race as to who cancels first. I am thinking of typical errors such as:
- Stream gets deleted
- Stream Consumer get deleted
- Credentials are invalid